main.py

```diff
@@ -137,26 +140,79 @@ def _validate_validation_status(response: requests.Response) -> bool:
         sys.exit(1)
 
 
-def download_html_report(token: str, request_id: str, payload: Dict[str, Any]):
-    response = _retry_request(
+def download_report(token: str, request_id: str, payload: Dict[str, Any], response_type: str):
+    response_types = {"html": "text/html",
+                      "json": "application/json"}
+
+    return _retry_request(
         "GET",
         f"https://appinspect.splunk.com/v1/app/report/{request_id}",
-        headers={"Authorization": f"bearer {token}", "Content-Type": "text/html"},
+        headers={"Authorization": f"bearer {token}", "Content-Type": response_types[response_type]},
         data=payload,
     )
 
-    with open("AppInspect_response.html", "w") as f:
+
+def download_and_save_html_report(token: str, request_id: str, payload: Dict[str, Any]):
+    response = download_report(token, request_id, payload, "html")
+
+    with open(f"AppInspect_response.html", "w") as f:
         f.write(response.text)
 
 
+def read_bytes_response_as_dict(response):
+    temp = response.content
+
+    return json.loads(temp.decode("utf-8"))
+
+
+def get_appinspect_failures_list(response_dict):
+    reports = response_dict['reports']
+    groups = reports[0]['groups']
+
+    failed_tests_list = []
+
+    for group in groups:
+        print(f"Now searching group {group['name']} for failed checks")
+        for check in group['checks']:
+            if check['result'] == "failure":
+                failed_tests_list.append(check["name"])
+
+    return failed_tests_list
+
+
+def read_yaml_as_dict(filename):
+    with open(f"./{filename}", 'r') as file:
+        try:
+            return yaml.safe_load(file)
+        except yaml.YAMLError as e:
+            print(e)
+
+
+def parse_except_dict_to_list(input_dict: Dict[str, str]):
+    expected_failures = []
+
+    for key, value in input_dict.items():
+        expected_failures.append(key)
+
+    return expected_failures
+
+
+def compare_failures(failures, expected):
+    if sorted(failures) != sorted(expected):
+        print("Appinspect failures doesn't match appinspect.expect file")
+        sys.exit(1)
+
+
 def parse_results(results):
     print(results)
     print(f"\n======== AppInspect Api Results ========")
     for metric, count  in results["info"].items():
         print("{0:>15}    :    {1: <4}".format(metric, count))
     if results["info"]["error"] > 0 or results["info"]["failure"] > 0:
         print("Error or failures in App Inspect")
-        sys.exit(1)
+        return True
+    else:
+        return False
```

Comment:
I am not sure if this implementation is a good practice @artemrys

Reply:
You can raise an exception if the results are not OK

Reply:
Done